### PR TITLE
61 add validation that flex request and flex offer are in the same conversation

### PR DESCRIPTION
**Pull request by eelcodenheijer**
* Added `conversationId` to `MessageSupport.getPreviousMessage()`
* Updated many validations that use this method
* Updated tests

### CODE REVIEW

File: core/src/main/java/org/lfenergy/shapeshifter/core/model/UftpMessage.java [hunk 1]
```diff
@@ -30,8 +30,11 @@ public static <T extends PayloadMessageType> UftpMessage<T> createOutgoing(UftpP
    * @param <U> The type of message that is being referenced (e.g. FlexRequest).
    * @return Reference to the referenced message.
    */
-  public <U extends PayloadMessageType> UftpMessageReference<U> referenceToPreviousMessage(String referencedMessageID, Class<U> referencedType) {
+  public <U extends PayloadMessageType> UftpMessageReference<U> findReferenceMessageInConversation(String referencedMessageID,
```
Comment (tomwetjens) [inline core/src/main/java/org/lfenergy/shapeshifter/core/model/UftpMessage.java:33]: The verb "find" implies (at least to me) that this method will actually start finding it, but actually it only creates a UftpMessageReference value object to hold the relevant reference values. So I think this method should be renamed to something no implying that it starts finding something.

File: core/src/test/java/org/lfenergy/shapeshifter/core/service/validation/base/ReferencedFlexRequestMessageIdValidatorTest.java
Comment (tomwetjens) [inline core/src/test/java/org/lfenergy/shapeshifter/core/service/validation/base/ReferencedFlexRequestMessageIdValidatorTest.java:1]: I think you should also modify ReferencedRequestMessageIdInResponseValidator to take the ConversationID into account

File: core/src/main/java/org/lfenergy/shapeshifter/core/model/UftpMessageReference.java
Comment (tomwetjens) [inline core/src/main/java/org/lfenergy/shapeshifter/core/model/UftpMessageReference.java:1]: I think all methods in UftpMessageSupport should take ConversationID account, so not only
`Optional<T> getPreviousMessage(UftpMessageReference<T> reference)`

but also:
`Optional<PayloadMessageType> getPreviousMessage(String messageID, String recipientDomain)`
and 
`boolean isValidOrderReference(String orderReference, String recipientDomain)`
and
`boolean existsFlexRevocation(String flexOfferMessageId, String recipientDomain)`

File: core/src/main/java/org/lfenergy/shapeshifter/core/model/UftpMessage.java (hunk 1 above)
Comment (MartenMeijboom98) [reply core/src/main/java/org/lfenergy/shapeshifter/core/model/UftpMessage.java:33]: `findMessageReferenceInConversation`?

File: core/src/main/java/org/lfenergy/shapeshifter/core/model/UftpMessageReference.java
Comment (eelcodenheijer) [reply core/src/main/java/org/lfenergy/shapeshifter/core/model/UftpMessageReference.java:1]: I've added `conversationId` to all methods of `UftpMessageSupport`

File: core/src/test/java/org/lfenergy/shapeshifter/core/service/validation/base/ReferencedFlexRequestMessageIdValidatorTest.java
Comment (eelcodenheijer) [reply core/src/test/java/org/lfenergy/shapeshifter/core/service/validation/base/ReferencedFlexRequestMessageIdValidatorTest.java:1]: Updated

File: core/src/main/java/org/lfenergy/shapeshifter/core/model/UftpMessage.java (hunk 1 above)
Comment (eelcodenheijer) [reply core/src/main/java/org/lfenergy/shapeshifter/core/model/UftpMessage.java]: i've reverted it to `referenceToPreviousMessage`

File: core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java [hunk 2]
```diff
@@ -25,38 +25,42 @@
...
-   * Gets a previously sent or received message by reference (usually for validation).
+   * Gets a previously sent or received message by conversation ID and reference (usually for validation).
    *
+   * @param conversationId The conversation ID
    * @param <T> The type of message.
    * @param reference The reference to the previous message.
    * @return The message that was either sent or received previously.
    */
-  <T extends PayloadMessageType> Optional<T> getPreviousMessage(UftpMessageReference<T> reference);
+  <T extends PayloadMessageType> Optional<T> getPreviousMessage(String conversationId, UftpMessageReference<T> reference);
```
Comment (tomwetjens) [inline core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java:45]: UftpMessageReference already contains the conversationID so I think this parameter is now duplicate?

File: core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java [hunk 3]
```diff
@@ -25,38 +25,42 @@
...
-   * Gets a previously received message by messageID (if any).
+   * Gets a previously received message by messageID (if any) and conversation ID.
    *
    * @param messageID The message ID.
+   * @param conversationId The conversation ID
    * @param recipientDomain The recipient domain.
    * @return The message that was received previously.
    */
-  Optional<PayloadMessageType> getPreviousMessage(String messageID, String recipientDomain);
+  Optional<PayloadMessageType> getPreviousMessage(String messageID, String conversationId, String recipientDomain);
```
Comment (tomwetjens) [inline core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java:35]: Naming of Id vs ID is now inconsistent

File: core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java [hunk 4]
```diff
@@ -25,38 +25,42 @@
...
-   * Checks whether a given order reference is present for a given recipient domain
+   * Checks whether a given order reference is present for a given conversation ID and recipient domain
    *
    * @param orderReference The order reference to be checked
+   * @param conversationId The conversation ID
    * @param recipientDomain The recipient domain for which the order reference should be checked
    * @return Whether the given order reference is present
    */
-  boolean isValidOrderReference(String orderReference, String recipientDomain);
+  boolean isValidOrderReference(String orderReference, String conversationId, String recipientDomain);
```
Comment (tomwetjens) [inline core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java:55]: For full correctness this also needs the flexOfferMessageId I think

File: core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java (hunk 3 above)
Comment (tomwetjens) [inline core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java:35]: After further thinking: this method is used to check for duplicate MessageIds which must be unique within a sender not within a conversation, so I think actually conversationID should not be there but senderDomain should be added.

File: core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java [hunk 5]
```diff
@@ -25,38 +25,42 @@
...
-   * Checks whether a flex revocation is present for a given flexOffer message and a recipient domain
+   * Checks whether a flex revocation is present for a conversation ID and given flexOffer message and recipient domain
    *
+   * @param conversationId The conversation ID
    * @param flexOfferMessageId The flexOffer message ID
    * @param recipientDomain The recipient domain
    * @return Whether a flex revocation is present
    */
-  boolean existsFlexRevocation(String flexOfferMessageId, String recipientDomain);
+  boolean existsFlexRevocation(String conversationId, String flexOfferMessageId, String recipientDomain);
```
Comment (tomwetjens) [inline core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java:65]: Discussed: maybe change this to "Optional<FlexRevocation> findFlexRevocation(String conversationId, flexOfferMessageID, senderDomain, recipientDomain)"

Also there is a missing validation: FlexRevocation may only be sent by the senderDomain that originally sent the FlexOffer

File: core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java (hunk 2 above)
Comment (eelcodenheijer) [reply core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java]: Changed to `findReferencedMessage`

File: core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java (hunk 4 above)
Comment (eelcodenheijer) [reply core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java]: I added the validation `findReferencedMessage(FlexSettlementMessageID)` in a separate new validator `ReferencedFlexSettlementReferenceValidator`

File: core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java (hunk 3 above)
Comment (eelcodenheijer) [reply core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java]: changed to `findDuplicateMessage`

File: core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java (hunk 5 above)
Comment (eelcodenheijer) [reply core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java]: Changed to
`Optional<FlexOfferRevocation> findFlexRevocation(String conversationID, String flexOfferMessageID, String senderDomain, String recipientDomain);`

File: core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java [hunk 6]
```diff
@@ -25,38 +27,42 @@
...
-   * Checks whether a given order reference is present for a given recipient domain
+   * Checks whether a given order reference is present for a given conversation ID and recipient domain
    *
    * @param orderReference The order reference to be checked
+   * @param conversationID The conversation ID
    * @param recipientDomain The recipient domain for which the order reference should be checked
    * @return Whether the given order reference is present
    */
-  boolean isValidOrderReference(String orderReference, String recipientDomain);
+  boolean isValidOrderReference(String orderReference, String conversationID, String recipientDomain);
```
Comment (tomwetjens) [inline core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java:56]: This method can be removed right?

Comment (eelcodenheijer) [reply core/src/main/java/org/lfenergy/shapeshifter/core/service/validation/UftpMessageSupport.java]: Yes, it has now been removed also

**Pull request merged**